server/utils.go

```diff
@@ -141,3 +144,119 @@ func isValidURL(rawURL string) error {
 
 	return nil
 }
+
+// filterLines filters lines in a string from start to end.
+func filterLines(s string, start, end int) (string, error) {
+	scanner := bufio.NewScanner(strings.NewReader(s))
+	var buf strings.Builder
+	for i := 1; scanner.Scan() && i <= end; i++ {
+		if i < start {
+			continue
+		}
+		buf.Write(scanner.Bytes())
+		buf.WriteByte(byte('\n'))
+	}
+
+	return buf.String(), scanner.Err()
+}
+
+// getLineNumbers returns the start and end lines from an anchor tag
+// of a GitLab permalink.
+func getLineNumbers(s string) (start, end int) {
+	// split till -
+	parts := strings.Split(s, "-")
+
+	if len(parts) > 2 {
+		return -1, -1
+	}
+
+	switch len(parts) {
+	case 1:
+		// just a single line
+		l := getLine(parts[0])
+		if l == -1 {
+			return -1, -1
+		}
+		if l < permalinkLineContext {
+			return 0, l + permalinkLineContext
+		}
+		return l - permalinkLineContext, l + permalinkLineContext
+	case 2:
+		// a line range
+		start := getLine(parts[0])
+		end := getLine(parts[1])
+		if start != -1 && end != -1 && start > end {
+			return -1, -1
+		}
+		return start, end
+	}
+	return -1, -1
+}
+
+// getLine returns the line number in int from a string
+// of form L<num>.
+func getLine(s string) int {
+	// check starting L and minimum length.
+	if !strings.HasPrefix(s, "L") || len(s) < 2 {
+		return -1
+	}
+
+	line, err := strconv.Atoi(s[1:])
+	if err != nil {
+		return -1
+	}
+	return line
+}
+
+// isInsideLink reports whether the given index in a string is preceded
+// by zero or more space, then (, then ].
+//
+// It is a poor man's version of checking markdown hyperlinks without
+// using a full-blown markdown parser. The idea is to quickly confirm
+// whether a permalink is inside a markdown link or not. Something like
+// "text ]( permalink" is rare enough. Even then, it is okay if
+// there are false positives, but there cannot be any false negatives.
+//
+// Note: it is fine to go one byte at a time instead of one rune because
+// we are anyways looking for ASCII chars.
+func isInsideLink(msg string, index int) bool {
```

Comment:
Is this implementation or explanation from another source?

Reply:
Yes, this implementation is taken from the Github plugin.

Reply:
Can we link to the code here using a GitHub permalink?

Reply:
Done